subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/WatchingVirtualFileSystem.java

```diff
@@ -175,28 +179,35 @@ public void beforeBuildFinished(boolean watchingEnabled) {
      */
     private synchronized void startWatching() {
         if (watchRegistry != null) {
+            if (!eventConsumerThread.isAlive()) {
+                throw new RuntimeException("The thread consuming the file events stopped for an unknown reason");
+            }
             return;
         }
         try {
             long startTime = System.currentTimeMillis();
             watchRegistry = watcherRegistryFactory.startWatcher(new FileWatcherRegistry.ChangeHandler() {
                 @Override
                 public void handleChange(FileWatcherRegistry.Type type, Path path) {
-                    try {
-                        fileEvents.put(new FileEvent(path, type));
-                    } catch (InterruptedException e) {
-                        Thread.currentThread().interrupt();
-                        throw new RuntimeException(e);
-                    }
+                    addToQueueOrReportOverflow(FileEvent.changed(path, type));
                 }
 
                 @Override
                 public void handleLostState() {
-                    try {
-                        fileEvents.put(new FileEvent(true));
-                    } catch (InterruptedException e) {
-                        Thread.currentThread().interrupt();
-                        throw new RuntimeException(e);
+                    addToQueueOrReportOverflow(FileEvent.lostState());
```

Comment:
This will try to add a lost state event, and if it fails, it will try to add another lost state event after clearing the queue. I think the shared code should be only about the lost state event that should always clear the queue.

Reply:
I am now always clearing the queue on lost state.